File: backend/app/services/stats_service.py
from datetime import date
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories import stats_repository


async def get_summary(
    db: AsyncSession,
    user_id: UUID,
    tz: str,
    start_date: date | None,
    end_date: date | None,
) -> dict:
    return await stats_repository.get_summary(db, user_id, tz, start_date, end_date)


async def get_trend(
    db: AsyncSession,
    user_id: UUID,
    tz: str,
    granularity: str,
    start_date: date | None,
    end_date: date | None,
) -> list[dict]:
    return await stats_repository.get_trend(db, user_id, tz, granularity, start_date, end_date)


async def get_heatmap(
    db: AsyncSession,
    user_id: UUID,
    tz: str,
    start_date: date | None,
    end_date: date | None,
) -> list[dict]:
    return await stats_repository.get_heatmap(db, user_id, tz, start_date, end_date)

File: backend/internal/service/stats_service.go
package service

import (
	"context"

	"angerlog/internal/models"
	"angerlog/internal/repository"

	"github.com/google/uuid"
)

type StatsService struct {
	repo *repository.StatsRepository
}

func NewStatsService(repo *repository.StatsRepository) *StatsService {
	return &StatsService{repo: repo}
}

func (s *StatsService) GetSummary(ctx context.Context, userID uuid.UUID, tz string, startDate, endDate *string) (*models.SummaryOut, *models.AppError) {
	summary, err := s.repo.GetSummary(ctx, userID, tz, startDate, endDate)
	if err != nil {
		return nil, models.ErrInternal
	}
	return summary, nil
}

func (s *StatsService) GetTrend(ctx context.Context, userID uuid.UUID, tz, granularity string, startDate, endDate *string) ([]models.TrendPoint, *models.AppError) {
	if granularity != "day" && granularity != "week" && granularity != "month" {
		return nil, models.ErrParam
	}

	trend, err := s.repo.GetTrend(ctx, userID, tz, granularity, startDate, endDate)
	if err != nil {
		return nil, models.ErrInternal
	}
	return trend, nil
}

func (s *StatsService) GetHeatmap(ctx context.Context, userID uuid.UUID, tz string, startDate, endDate *string) ([]models.HeatmapCell, *models.AppError) {
	heatmap, err := s.repo.GetHeatmap(ctx, userID, tz, startDate, endDate)
	if err != nil {
		return nil, models.ErrInternal
	}
	return heatmap, nil
}

